kong/plugins/http-log/handler.lua

```diff
@@ -18,68 +17,146 @@ HttpLogHandler.PRIORITY = 12
 HttpLogHandler.VERSION = "1.0.0"
 
 
-local buffers = {} -- buffers per-route
+local queues = {} -- queues per-route
 
 
--- Only provide `name` when deriving from this class. Not when initializing an instance.
-function HttpLogHandler:new(name)
-  name = name or "http-log"
-  HttpLogHandler.super.new(self, name)
+-- Parse host url.
+-- @param `url` host url
+-- @return `parsed_url` a table with host details:
+-- scheme, host, port, path, query, userinfo
+local function parse_url(host_url)
+  local parsed_url = url.parse(host_url)
+  if not parsed_url.port then
+    if parsed_url.scheme == "http" then
+      parsed_url.port = 80
+    elseif parsed_url.scheme == "https" then
+      parsed_url.port = 443
+    end
+  end
+  if not parsed_url.path then
+    parsed_url.path = "/"
+  end
+  return parsed_url
+end
 
-  self.ngx_log = ngx.log
---  self.ngx_log = function(lvl, ...)
---    ngx_log(lvl, "[", name, "] ", ...)
---  end
 
-  self.name = name
+-- Sends the provided payload (a string) to the configured plugin host
+-- @return true if everything was sent correctly, falsy if error
+-- @return error message if there was an error
+local function send_payload(self, conf, payload)
+  local method = conf.method
+  local timeout = conf.timeout
+  local keepalive = conf.keepalive
+  local content_type = conf.content_type
+  local http_endpoint = conf.http_endpoint
+
+  local ok, err
+  local parsed_url = parse_url(http_endpoint)
+  local host = parsed_url.host
+  local port = tonumber(parsed_url.port)
+
+  local httpc = http.new()
+  httpc:set_timeout(timeout)
+  ok, err = httpc:connect(host, port)
+  if not ok then
+    kong.log.err("failed to connect to ", host, ":", tostring(port), ": ", err)
+    return false
+  end
+
+  if parsed_url.scheme == "https" then
+    local _, err = httpc:ssl_handshake(true, host, false)
+    if err then
+      kong.log.err("failed to do SSL handshake with ",
+                   host, ":", tostring(port), ": ", err)
+      return false
+    end
+  end
+
+  local res, err = httpc:request({
+    method = method,
+    path = parsed_url.path,
+    query = parsed_url.query,
+    headers = {
+      ["Host"] = parsed_url.host,
+      ["Content-Type"] = content_type,
+      ["Content-Length"] = #payload,
+      ["Authorization"] = parsed_url.userinfo and (
+        "Basic " .. ngx_encode_base64(parsed_url.userinfo)
+      ),
+    },
+    body = payload,
+  })
+  if not res then
+    kong.log.err("failed request to ", host, ":", tostring(port), ": ", err)
+    httpc:set_keepalive(keepalive)
+    return false
+  end
+
+  -- read and discard response body
+  -- TODO should we fail if response status was >= 500 ?
```

Comment:
at least log an error in case >= 500 ?

Reply:
I have reworked this ( >= 400 instead of 500). 

The plugin will now consider it a "failure". An error will be logged and this function will return `false`, so the batch will be retried.

@hishamhm you might consider this one of those "slightly breaking changes" we have recently talked about (before, a 500 status would not trigger a retry).

Reply:
> before, a 500 status would not trigger a retry

I believe this behavior was never really documented, and it's not the kind of thing that users would "rely on", so I think it's fine to change.